src/core/message-parsing/handlers/BaseDirectiveHandler.ts

```diff
@@ -0,0 +1,28 @@
+import * as sax from "sax"
+import { DirectiveHandler } from "../DirectiveHandler"
+import { ParseContext } from "../ParseContext"
+import { TextDirective } from "../directives"
+
+export abstract class BaseDirectiveHandler implements DirectiveHandler {
+	abstract readonly tagName: string
+
+	canHandle(tagName: string): boolean {
```

Comment:
This could be a good place to separate tool-specific parameters. The base class could have something like an abstract `getToolParameterSchema` for which each subclass implementation would return something a validation structure specific to each tool. 

This would allow you to recursively validate the input as it arrives, and keep each tool parameter logic separated into the proper location specific to that tool, so they do not get mixed up as noted in the other comment:

```ts
read_file : {
	path : pathValidatorFnRef,
	start_line: "int",
	end_line: "int"
}
```

```ts
read_file : {
	args : {
		[
			file : {
				path : pathValidatorFnRef,
				line_range : [
					lineRangeValidatorFnRef
					]
			},
		]
	}
}
```

Reply:
> I agree, honestly never thought you guys would approve even this much change, So I actually was showing a small amount of restraint. albeit not much ;)

Changes like this are necessarily big, and I believe this change is definitely necessary!